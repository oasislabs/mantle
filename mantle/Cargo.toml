[package]
name = "mantle"
version = "0.1.0"
license = "Apache-2.0"
description = "A humane programming model for blockchain."
edition = "2018"
repository = "https://github.com/oasislabs/mantle.git"
authors = ["Oasis Labs <feedback@oasislabs.com>"]
readme = "README.md"
keywords = ["mantle"]

[dependencies]
failure = "0.1"
mantle-macros = { version = "0.1", path = "../mantle-macros" }
mantle-types = { version = "0.1", path = "../mantle-types" }
serde = { version = "1.0", features = ["derive"] }
serde_cbor = "0.9"
tiny-keccak = "1.4"

[dev-dependencies]
mantle-test = { path = "../mantle-test" }
serde = { version = "1.0", features = ["derive"] }
tiny-keccak = "1.4"

[features]
platform-alloc = []
create2 = []
