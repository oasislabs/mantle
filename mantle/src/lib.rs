#![feature(bind_by_move_pattern_guards, linkage, non_exhaustive, trait_alias)]

extern crate mantle_macros;

pub mod error;
pub mod exe;
pub mod ext;

pub mod reexports {
    pub use serde;
    pub use serde_cbor;
    pub use tiny_keccak;
}

pub use mantle_macros::{Event, Service};
pub use mantle_types::Address;

pub use crate::exe::*;

/// This macro is used to define the "main" service.
///
/// ## Example

/// ```norun
/// fn main() {
///    mantle::service!(TheMainService);
/// }
/// ```
#[macro_export]
macro_rules! service {
    ($svc:path) => {};
}

pub trait AddressExt {
    fn transfer<'a>(&self, value: u64) -> Result<(), crate::error::Error>;

    fn balance(&self) -> u64;
}

impl AddressExt for Address {
    fn transfer<'a>(&self, value: u64) -> Result<(), crate::error::Error> {
        crate::ext::transfer(self, value)
    }

    fn balance(&self) -> u64 {
        crate::ext::balance(self).unwrap()
    }
}
