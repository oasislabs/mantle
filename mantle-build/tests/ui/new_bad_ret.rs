use mantle::Context;

#[derive(mantle::Service, Default)]
pub struct Counter(u32);

impl Counter {
    pub fn new(ctx: &Context, start_count: u32) -> () {
        Self(start_count);
    }
}

fn main() {
    mantle::service!(Counter);
}
