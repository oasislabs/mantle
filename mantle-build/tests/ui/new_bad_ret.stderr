error: Service constructor must return `Self` (aka `Counter`)
 --> $DIR/new_bad_ret.rs:7:52
  |
7 |     pub fn new(ctx: &Context, start_count: u32) -> Self {
  |                                                    ^^^^

error: aborting due to previous error

